### PR TITLE
New Parts, Microcontrollers, APIs, and Cleanup

### PR DESCRIPTION
**Pull request by m-hilgendorf**
This is a big PR, cherry picking the work done by @PhilippeFerreiraDeSousa, @d-haldane, and @electricfungi. Thanks everyone!

## New Checks
- `CrystalOscillator` property struct and associated checks
- Replaces properties in many components (abracon, epson, espressif, etc)

## New Components
-  abracon oscillators:
  - AMB7, AMPM series
- Diodes Incorporated AP[2112](https://www.youtube.com/watch?v=AZm1_jtY1SQ)
- Texas Intstruments 
  - SN65HVD1786
- ST Microelectronics
  - Many STM32 MCUs!

## New Bundles
- high-freq-oscillator
- I2S
- 4-pin JTAG

## New APIs:

- MCUs added to the dbquery API
- STM32 API for querying MCUs that solve the requirements and set up correct modules
- `make-default-board` now accepts arbitrary shapes instead of dimensions
- `add-mounting-holes` accepts shapes to place the mounting holes, and interface for selecting which holes to place

## Designs:
- designs updated for new APIs
- MCU example design

## Cleanup
- Additional inline documentation added to some function calls internally.

### CODE REVIEW

File: components/st-microelectronics/STM32F031F6P6-supports.stanza [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+; STM32L031F(4-6)Px
+
+public defn stm-support ( ) :
+  inside pcb-component :
+
+    supports swd:
+      swd.swdio  => self.PA[13]
+      swd.swdclk => self.PA[14]
+
+    supports gpio:
```
Comment (bpellenard) [inline components/st-microelectronics/STM32F031F6P6-supports.stanza:22]: is there a way to have a for loop for all these supports gpio?

File: components/st-microelectronics/STM32F031F6P6-supports.stanza [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      gpio.gpio => self.PA[9]
+
+    supports gpio:
+      gpio.gpio => self.PA[10]
+
+    supports gpio:
+      gpio.gpio => self.PA[13]
+
+    supports gpio:
+      gpio.gpio => self.PA[14]
```
Comment (bpellenard) [inline components/st-microelectronics/STM32F031F6P6-supports.stanza:65]: is there a way to have a for loop for all these supports gpio?

**Comment by bpellenard** [comment]
Is there a way to split into multiple PRs? One for the new components, one for the new bundles, one for the checks for example

File: components/st-microelectronics/STM32F031F6P6-supports.stanza (hunk 2 above)
Comment (m-hilgendorf) [reply components/st-microelectronics/STM32F031F6P6-supports.stanza:65]: No. This is autogenerated code.

File: components/nordic/nRF52840.stanza [hunk 3]
```diff
@@ -183,41 +187,70 @@ public pcb-component component :
...
+public pcb-module module :
+  port usb : usb-2
+  pin gnd
+  inst mcu : ocdb/nordic/nRF52840/component
+
+  net (gnd, usb.vbus.gnd, mcu.VSS, mcu.VSS_PA)
+
+  ; Configure to be in normal voltage mode, DCDC1 enabled (config 5 from datasheet) Improvement: make power parametric
+  ; Bypass caps
+  bypass-cap-strap(mcu.DEC1, gnd, 100.0e-9)
```
[the remark this answers is not in the document]
Comment (d-haldane) [reply components/nordic/nRF52840.stanza:210]: These caps and values are design specific and can't be hardcoded as properties in the component

**Comment by bpellenard** [comment]
Is it ready to be merged?

**Comment by d-haldane** [comment]
It is not ready to be merged yet

**Review by PhilippeFerreiraDeSousa — APPROVED** [review]
Needed by Cobalt by Monday

File: components/future-designs/FT232RL.stanza [hunk 4]
```diff
@@ -80,8 +80,8 @@ public pcb-component component:
     val vol-min = PWL([[1.8 0.06] [2.8 0.3] [3.3 0.3] [5.0 0.3]])[vdd]
     val vih = 1.5
     val vil = vdd * 0.3 ;guess
-    val drive-props = DigitalOutput(CMOSOutput(Interval(vol-min, vol-max, false), Interval(voh-min, voh-max, false)))
-    val input-props = DigitalInput(vil, vih)
+    val drive-props = DigitalOutput(CMOSOutput(Interval(vol-min, vol-max, false), Interval(voh-min, voh-max, false)), false, VCCIO, GND)
+    val input-props = DigitalInput(vil, vih, VCCIO, GND, 1.0e-6) ; TODO: Leakage current is a guess
```
Comment (bpellenard) [inline components/future-designs/FT232RL.stanza:84]: do you want to address this TODO later?

Comment (m-hilgendorf) [reply components/future-designs/FT232RL.stanza:84]: yes. This PR is already merged.

File: components/raytac/MDBT50Q.stanza [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  val pady = 0.4
+  val pitch = 0.8
+  val offset = 0.9
+
+  pad p[1] : smd-pad(padx, pady) at loc(0.6, 11.5)
+  pad p[2] : smd-pad(padx, pady) at loc(0.6, 10.4)
+  pad p[3] : smd-pad(padx, pady) at loc(0.6, 9.6)
+  for i in 4 to 15 do :
+    val div = to-double(i % 2)
+    if div > 0.0 :
```
Comment (bpellenard) [inline components/raytac/MDBT50Q.stanza:102]: I do not understand the logic here

File: components/raytac/MDBT50Q.stanza [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+      pad p[i] : smd-pad(padx, pady) at loc(0.6 + offset, 8.0 - ((to-double((i - 5) / 2) + 0.5) * pitch))
+    else :
+      pad p[i] : smd-pad(padx, pady) at loc(0.6, 8.0 - (to-double((i - 4) / 2) * pitch))
+
+  for i in 15 to 34 do :
+    if i >= 19 and i <=29 :
+      val div = to-double(i % 2)
+      if div > 0.0 :
```
Comment (bpellenard) [inline components/raytac/MDBT50Q.stanza:110]: likewise

**Pull request merged**